include/flang/Lower/PFTBuilder.h

```diff
@@ -168,102 +129,114 @@ struct Evaluation {
+  // to statement nodes.  One exception is that an internal construct node is
+  // a convenient place for a constructExit link that applies to exits from any
+  // statement within the construct.  The controlSuccessor member is used for
+  // nonlexical successors, such as linking to a GOTO target.  For multiway
+  // branches, controlSuccessor is set to one of the targets (might as well be
+  // the first target).  Successor and exit links always target statements.
+  //
+  // An unstructured construct is one that contains some form of goto.  This
+  // is indicated by the isUnstructured member flag, which may be set on a
+  // statement and propagated to enclosing constructs.  This distinction allows
+  // a structured IF or DO statement to be materialized with custom structured
+  // FIR operations.  An unstructured statement is materialized as mlir
+  // operation sequences that include explicit branches.
+  //
+  // There are two mlir::Block members.  The block member is set for statements
+  // that begin a new block.  If a statement may have more than one associated
+  // block, this member must be the block that would be the target of a branch
+  // to the statement.  The prime example of a statement that may have multiple
+  // associated blocks is NonLabelDoStmt, which may have a loop preheader block
+  // for loop initialization code, and always has a header block that is the
+  // target of the loop back edge.  If the NonLabelDoStmt is a concurrent loop,
+  // there may be an arbitrary number of nested preheader, header, and mask
+  // blocks.  Any such additional blocks in the localBlocks member are local
+  // to a construct and cannot be the target of an unstructured branch.  For
+  // NonLabelDoStmt, the block member designates the preheader block, which may
+  // be absent if loop initialization code may be appended to a predecessor
+  // block.  The primary loop header block is localBlocks[0], with additional
+  // DO CONCURRENT blocks at localBlocks[1], etc.
+  //
+  // The printIndex member is only set for statements.  It is used for dumps
+  // and does not affect FIR generation.  It may also be helpful for debugging.
+
+  EvaluationVariant u;
+  ParentType parentType;
+  parser::CharBlock position{};
+  std::optional<parser::Label> label{};
+  std::unique_ptr<EvaluationList> evaluationList; // nested evaluations
+  Evaluation *parentConstruct{nullptr};  // set for nodes below the top level
+  Evaluation *lexicalSuccessor{nullptr}; // set for ActionStmt, ConstructStmt
+  Evaluation *controlSuccessor{nullptr}; // set for some statements
+  Evaluation *constructExit{nullptr};    // set for constructs
+  bool isNewBlock{false};                // evaluation begins a new basic block
+  bool isUnstructured{false};        // evaluation has unstructured control flow
+  bool skip{false};                  // evaluation has been processed in advance
+  class mlir::Block *block{nullptr}; // isNewBlock block
```

Comment:
Many of these properties appear to have purposes that are not applicable to a general `Evaluation` (computational code) but are rather specific to those cases where computational code implies a transfer of control. For example, most evals will perform some computation (of varying complexity) and fall through to the next eval. (Linear order property.) Every one of these evals will now have, to name a couple, a `controlSuccessor` and a `constructExit`.

We should try to avoid bundling every statement of the Fortran input in the PFT with a bag of state that is specialized to a small subset of statements.

Reply:
I think it is best to view the PFT as a data structure that includes branch links, not just the core tree structure per se.  I see no benefit in attempting to distinguish between "tree links" and "cross links".  Just as some leaf evaluations have no leaf predecessor or successor,  and the root evaluation has no parent, some have no control successor, but others do.

Reply:
Can we use subclassing?

Reply:
Having only a vague idea of what that might mean, I don't see any value to that.  (It seems like added complexity for little benefit.)  Having a better idea might or might not change that.  Is space an issue?  These members are fairly small.

Reply:
Oh, that reminds me.  Naming the root of a PFT as "program" or "pgm" is misleading, because at most a single such root in a large program will actually be a PROGRAM.  Something like compilationUnit might be better.  But this never seemed important enough to consider changing.  (And it will most likely remain below my radar in the future as well, unless someone else brings it up.)

Reply:
Hmm. I think I used "program" in a Fortran jargon sense of "program unit", not to be confused with the narrower case of "main program".

Reply:
WRT, subclassing. I'm suspicious that Evaluation could get really out of hand if we want to annotate different Evaluation nodes in different ways and just turn the Evaluation class into the union of all side information that lowering might ever need for any node in the tree. At this point, it doesn't matter because we only have the CFG use cases. So we can table this until it becomes an issue.